### PR TITLE
feat(netlify): experimental support v2 functions ISR via targeted cache-control headers

### PR DESCRIPTION
**Pull request by serhalp**
### 🔗 Linked issue

Supersedes https://github.com/unjs/nitro/pull/1823.
Closes https://github.com/unjs/nitro/issues/1842.
Closes https://github.com/unjs/nitro/issues/1938.

### ❓ Type of change

- [x] 📖 Documentation (updates to the documentation, readme, or JSdoc annotations)
- [x] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

Netlify [introduced a new Functions interface last year](https://www.netlify.com/blog/introducing-netlify-functions-2-0/). This updates Nitro to use this interface.

Previously, ISR in Nitro had been implemented on Netlify via its proprietary [On-Demand Builders](https://docs.netlify.com/configure-builds/on-demand-builders/) platform primitive, which only supports "v1" Functions. Now that there's momentum around [Targeted HTTP Cache Control](https://httpwg.org/specs/rfc9213.html) and that [Netlify supports it](https://docs.netlify.com/platform/caching/#supported-cache-control-headers), we can implement ISR in Nitro via these simple, standard HTTP headers instead.

By virtue of switching to the standard Netlify Functions interface, this also happens to:
- allow streaming responses
- add better support for "hybrid rendering", e.g. if you set `prerender: true` on a specific route, that will actually work as intended now
- remove the imposed 60s delay after a background generation request before which the new response will be served for subsequent requests (this was an ODB limitation)
- fix a bug where we were setting `set-cookie: undefined` on responses
- remove usage of an undocumented `functions-internal/server.json` Netlify config file
- ... which wasn't working when using `netlify dev` in local development anyway

This also configures a human-friendly name and some metadata on the function.

Note: I removed the `netlify_builder` preset, since there's no reason to use this anymore. I'm not even sure why it was ever exposed as a preset. I did this as a separate commit, so if you consider this a breaking change I can pull it out into a separate PR.

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: src/presets/index.ts [hunk 1]
```diff
@@ -14,7 +14,7 @@ export { digitalOcean } from "./digital-ocean";
 export { firebase } from "./firebase";
 export { heroku } from "./heroku";
 export { edgio } from "./edgio";
-export { netlify, netlifyBuilder, netlifyEdge, netlifyStatic } from "./netlify";
+export { netlify, netlifyEdge, netlifyStatic } from "./netlify";
```
Comment (pi0) [inline src/presets/index.ts:17]: We need to keep `netlifyBuilder` for backward compatibility for users who might be using it currently

File: src/presets/netlify.ts [hunk 2]
```diff
@@ -4,17 +4,32 @@ import { defineNitroPreset } from "../preset";
...
+// This is written to the functions directory. It just re-exports the compiled handler,
+// along with its config. We do this instead of compiling the entrypoint directly because
+// the Netlify platform actually statically analyzes the function file to read the config;
+// if we compiled the entrypoint directly, it would be chunked and wouldn't be analyzable.
+const generateNetlifyFunction = (handlerRelativePath: string) => {
+  return /* js */ `
+    export { default } from "${handlerRelativePath}";
+    export const config = {
+      name: "Nitro server handler",
+      generator: "${nitroPkg.name}@${nitroPkg.version}",
```
Comment (pi0) [inline src/presets/netlify.ts:16]: I think we can refactor this function inside hook or pass `nitro` instance to use precise generator/framework info from `nitro.options.framework`

File: src/runtime/entries/netlify.ts
Comment (pi0) [inline src/runtime/entries/netlify.ts:1]: I think we can introduce changes to new entries `netlify-v1` and `netlify-v2`

File: test/tests.ts [hunk 3]
```diff
@@ -81,7 +81,7 @@ export async function setupTest(
       "vercel-edge",
       "winterjs",
     ].includes(preset),
-    isLambda: ["netlify", "aws-lambda"].includes(preset),
+    isLambda: ["aws-lambda"].includes(preset),
```
Comment (pi0) [inline test/tests.ts:84]: Netlify functions are not lambda anymore? or what test it fixes?

File: src/presets/netlify.ts [hunk 4]
```diff
@@ -37,27 +57,6 @@ export const netlify = defineNitroPreset({
           "utf8"
         );
       }
-
-      const functionConfig = {
-        config: { nodeModuleFormat: "esm" },
-        version: 1,
-      };
```
Comment (pi0) [inline src/presets/netlify.ts:44]: Is v2 fully compatible with v1 for seamless migration of users?

I would prefer to keep v1 and incrementally add v2 implementation to allow opt-in for all new users unless netlify platform is not supporting v1 format anymore

**Comment by pi0** [comment]
(omitted 2fc17ffcf546d33d54c4afec343c3513723c2e19 for now to reduce diff)

File: src/runtime/entries/netlify.ts [hunk 5]
```diff
@@ -1,28 +1,64 @@
...
+  const r = await nitroApp.localCall({
+    url: relativeUrl,
+    headers: req.headers,
+    method: req.method,
+    body: req.body,
+  });
+
+  const headers = normalizeResponseHeaders({
+    ...r.headers,
+    ...getCacheHeaders(relativeUrl),
```
Comment (Rigo-m) [inline src/runtime/entries/netlify.ts]: Is it correct that getCacheHeaders gets spread after localCall headers?

**Comment by Rigo-m** [comment]
Super @serhalp 😍

**Review by pi0 — APPROVED** [review]
Thanks! Merging as is mainly to allow experimenting on main branch and do pre-release refactors. 

The confusing v1/v2 scheme that I added to list commits will be removed likely before releasing next versions.

**Pull request merged**